Infer output shape of Expand op

shinh opened the pull request.

shinh commented:
@houseroad : Can I request you to review this PR since you have reviewed a few PRs I made for shape inference?

houseroad commented:
Also could you please add a test case to https://github.com/onnx/onnx/blob/master/onnx/test/shape_inference_test.py

shinh commented:
Ah, I thought this op is `np.broadcast_to`. Now it handles mutlidirectional broadcasting. I've added tests for them, too. Thanks for your review!

shinh commented:
Ping. Could you take another look?

houseroad commented on onnx/defs/math/defs.cc:
```diff
@@ -1085,7 +1085,50 @@ ONNX_OPERATOR_SET_SCHEMA(
+          if (!target_shape_initializer) {
+            return;
+          }
+          // The targetShape vector represents the specified shape for output.
```
nit: targetShape => target_shape

shinh replied:
Done

houseroad commented on onnx/defs/math/defs.cc:
```diff
@@ -1085,7 +1085,50 @@ ONNX_OPERATOR_SET_SCHEMA(
+              ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+          for (int i = 0; i < input_shape.dim_size(); ++i) {
+            auto* new_dim = output_shape->add_dim();
+            int64_t target_dim = static_cast<size_t>(i) < target_shape.size() ? target_shape[i] : 1;
```
Dimensions are right aligned. Please check the case here: https://github.com/onnx/onnx/blob/master/docs/Operators.md#expand

shinh replied:
Oops, sorry for the incorrect understanding. Done

houseroad commented on onnx/test/shape_inference_test.py:
```diff
@@ -234,6 +234,33 @@ def test_reshape_static_shape_constant(self):  # type: () -> None
+            [make_node("Expand", ['x', 'shape'], ['y'])],
+            [],
+            initializer=[make_tensor('shape', TensorProto.INT64, (2,), (1, 4))])
+        self._assert_inferred(graph, [make_tensor_value_info('y', TensorProto.UINT8, (2, 4, 3))])
```
this is an incorrect case.

shinh replied:
Done

houseroad commented on onnx/test/shape_inference_test.py:
```diff
@@ -234,6 +234,33 @@ def test_reshape_static_shape_constant(self):  # type: () -> None
+            [make_node("Expand", ['x', 'shape'], ['y'])],
+            [],
+            initializer=[make_tensor('shape', TensorProto.INT64, (2,), (2, 1))])
+        self._assert_inferred(graph, [make_tensor_value_info('y', TensorProto.UINT8, (2, 'd2', 'd3'))])
```
this should be d1, 2, d3

shinh replied:
Done

shinh commented:
> please add a test in which shape's dim is larger than input's dim.

Sorry, I missed this comment. Added a test and fixed the implementation. I hope I'm not missing anything now..

shinh commented:
Ping? Are there still issues?

shinh commented:
@houseroad friendly reminder! Is there still any issue?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/onnx/onnx?pullRequest=1855) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/onnx/onnx?pullRequest=1855) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/onnx/onnx?pullRequest=1855) it.</sub>

shinh commented:
It seems this was fixed by https://github.com/onnx/onnx/pull/2041
